pkg/querier/matrix.go

```diff
@@ -9,22 +9,29 @@ import (
 	"github.com/prometheus/common/model"
 	"github.com/prometheus/prometheus/tsdb/chunkenc"
 
+	"github.com/grafana/mimir/pkg/mimirpb"
 	"github.com/grafana/mimir/pkg/storage/chunk"
 	"github.com/grafana/mimir/pkg/storage/series"
-	"github.com/grafana/mimir/pkg/util"
+	"github.com/grafana/mimir/pkg/util/modelutil"
 )
 
 func mergeChunks(chunks []chunk.Chunk, from, through model.Time) chunkenc.Iterator {
 	samples := make([][]model.SamplePair, 0, len(chunks))
+	histograms := make([][]mimirpb.Histogram, 0, len(chunks))
 	for _, c := range chunks {
-		ss, err := c.Samples(from, through)
+		sf, sh, err := c.Samples(from, through)
 		if err != nil {
 			return series.NewErrIterator(err)
 		}
-
-		samples = append(samples, ss)
+		if len(sf) > 0 {
+			samples = append(samples, sf)
+		}
+		if len(sh) > 0 {
+			histograms = append(histograms, sh)
+		}
 	}
+	mergedSamples := modelutil.MergeNSampleSets(samples...)
+	mergedHistograms := modelutil.MergeNHistogramSets(histograms...)
```

Comment:
I don't think there's too much to gain, on the other hand it should be easy to do for floats since you just copy in the old implementation. And one could very simply replicate for histograms so go for it!

Reply:
You are right. I assumed the implementation of below functions without actually looking into it. We may just avoid some allocations, so will just do that.